codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -66,69 +70,135 @@
-     * @param bootstrap bootstrap used to instantiate child channels for remotely-created streams.
      */
-    public Http2MultiplexCodec(boolean server, Http2StreamChannelBootstrap bootstrap) {
-        if (bootstrap.parentChannel() != null) {
-            throw new IllegalStateException("The parent channel must not be set on the bootstrap.");
-        }
+    public Http2MultiplexCodec(boolean server, ChannelHandler inboundStreamHandler) {
         this.server = server;
-        this.bootstrap = new Http2StreamChannelBootstrap(bootstrap);
+        this.inboundStreamHandler = checkSharable(checkNotNull(inboundStreamHandler, "inboundStreamHandler"));
     }
 
-    @Override
-    public void handlerAdded(ChannelHandlerContext ctx) {
-        this.ctx = ctx;
-        bootstrap.parentChannel(ctx.channel());
+    private static ChannelHandler checkSharable(ChannelHandler handler) {
+        if ((handler instanceof ChannelHandlerAdapter && !((ChannelHandlerAdapter) handler).isSharable()) ||
+                !handler.getClass().isAnnotationPresent(Sharable.class)) {
+            throw new IllegalArgumentException("The handler must be Sharable");
+        }
+        return handler;
     }
 
     @Override
-    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
-        if (!(cause instanceof StreamException)) {
-            ctx.fireExceptionCaught(cause);
-            return;
+    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+        if (ctx.executor() != ctx.channel().eventLoop()) {
+            throw new IllegalStateException("EventExecutor must be EventLoop of Channel");
         }
+        this.ctx = ctx;
+        super.handlerAdded(ctx);
+    }
 
-        StreamException streamEx = (StreamException) cause;
-        try {
-            Http2StreamChannel childChannel = childChannels.get(streamEx.streamId());
-            if (childChannel != null) {
-                childChannel.pipeline().fireExceptionCaught(streamEx);
-            } else {
-                logger.warn(format("Exception caught for unknown HTTP/2 stream '%d'", streamEx.streamId()),
-                            streamEx);
-            }
-        } finally {
-            onStreamClosed(streamEx.streamId());
-        }
+    @Override
+    public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
+        channels.clear();
+        channelsToFireChildReadComplete.clear();
+        super.handlerRemoved(ctx);
     }
 
-    // Override this to signal it will never throw an exception.
     @Override
-    public void flush(ChannelHandlerContext ctx) {
-        ctx.flush();
+    public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
+        if (evt instanceof Http2FrameStreamEvent) {
+            Http2FrameStreamEvent streamEvt = (Http2FrameStreamEvent) evt;
+            switch (streamEvt.state()) {
+                case CLOSED:
+                    onStreamClosed(streamEvt.stream());
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)